src/mpid/ch4/src/ch4r_recvq.h

```diff
@@ -108,6 +112,7 @@ MPL_STATIC_INLINE_PREFIX MPIR_Request *MPIDIG_dequeue_unexp(int rank, int tag,
         req = curr->request;
         if (MPIDIG_match_unexp(rank, tag, context_id, req)) {
             DL_DELETE(*list, curr);
+            MPIR_T_event_instance(unexp_message_indices[1], MPI_T_CB_REQUIRE_MPI_RESTRICTED, &tag);
```

Comment:
I see. You want to avoid the function call overhead altogether. Only `MPIR_T_even_instance` is relevant to the performance concern, right? I would rather use another macro wrapper, that defined to empty when events are disabled, than the current mechanism, which seems more mysterious to me.

Reply:
So something like?
```
MPIR_T_event_instance (...) \
    do {} while (0)
```

Reply:
I am more thinking about using a separate macro, e.g.
```
#ifdef HAVE_MPIT_EVENTS
#define MPIR_T_DO_EVENT MPIR_T_event_instance(...)
#else
#define MPIR_T_DO_EVENT do{}while(0)
#endif
```

`MPIR_T_event_instance` is a regular function, so reader may miss that it potentially is defined away. With a different macro name, the difference is explicit, and the reader can quickly check and understand the logics behind it.

Reply:
Ah, yes. That's more like what the PVAR code does. I can make that change. `(...)` is from C99 onwards: https://en.cppreference.com/w/c/preprocessor/replace

Reply:
Added MPIR_T_DO_EVENT.